File: 599/MinIndexSumOfTwoLists.py
import sys

class Solution(object):
    def findRestaurant(self, list1, list2):
        """
        :type list1: List[str]
        :type list2: List[str]
        :rtype: List[str]
        """
        s = set(list1) & set(list2)
        sums = {e: list1.index(e) + list2.index(e) for e in s}
        minsum = reduce(lambda a,b: a if a < b else b, sums.values())
        ret = filter(lambda x: sums[x] <= minsum, sums.keys())
        return ret

File: 599/MinIndexSumOfTwoLists.java
import java.util.Map;
import java.util.HashMap;
import java.util.Set;
import java.util.HashSet;
import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;

class MinIndexSumOfTwoLists {
    public String[] findRestaurant(String[] list1, String[] list2) {
        List<String> l1 = new ArrayList<>(Arrays.asList(list1));
        List<String> l2 = new ArrayList<>(Arrays.asList(list2));
        Set<String> s1 = new HashSet<>();
        Set<String> s2 = new HashSet<>();
        Set<String> s = new HashSet<>();
        Map<String, Integer> m = new HashMap<>();
        
        /* get the intersection set */
        for (String x: l1) {
            s1.add(x);
        }
        for (String x: l2) {
            s2.add(x);
        }
        for (String x: s1) {
            if (s2.contains(x)) {
                s.add(x);
            }
        }
        
        /* get minimum sum of index and build the sum map*/
        int minsum = 2000;
        for (String x: s) {
            int sum = l1.indexOf(x) + l2.indexOf(x);
            m.put(x, sum);
            if (sum < minsum) minsum = sum;
        }
        
        /* get all string that has sum equal to min sum */
        List<String> ret = new ArrayList<>();
        for (String x: m.keySet()) {
            if (m.get(x) == minsum) {
                ret.add(x);
            }
        }
        
        return ret.toArray(new String[ret.size()]);
    }
}
